### PR TITLE
[MXNET-580] Add SN-GAN example

### PR DESCRIPTION
**Pull request by stu1130**
## Description ##
Add Spectral Normalization GAN example

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
Spectral Normalization GAN example

## Comments ##
N/A

### CODE REVIEW

File: example/gluon/sn-gan/spectral_normalization_gan.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+    def spectral_norm(self):
+        """ spectral normalization """
+        w = self.params.get('weight').data(CTX)
+        w_mat = w
+        w_mat = nd.reshape(w_mat, [w_mat.shape[0], -1])
+
+        _u = self.u.data(CTX)
+        _v = None
+
+        for _ in range(1):
```
Comment (vandanavk) [inline example/gluon/sn-gan/spectral_normalization_gan.py]: This for loop is executed only once? Do we need a loop in that case?

File: example/gluon/sn-gan/spectral_normalization_gan.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+            _v = nd.L2Normalization(nd.dot(_u, w_mat))
+            _u = nd.L2Normalization(nd.dot(_v, w_mat.T))
+
+        sigma = nd.sum(nd.dot(_u, w_mat) * _v)
+
+        self.params.setattr('u', _u)
+
+        return w / sigma
```
Comment (vandanavk) [inline example/gluon/sn-gan/spectral_normalization_gan.py]: Is there a chance that sigma would be 0?

File: example/gluon/sn-gan/spectral_normalization_gan.py (hunk 1 above)
Comment (stu1130) [reply example/gluon/sn-gan/spectral_normalization_gan.py]: The original paper use for loop to calculate the result. I should make it as a hyperparameter

File: example/gluon/sn-gan/spectral_normalization_gan.py (hunk 2 above)
Comment (stu1130) [reply example/gluon/sn-gan/spectral_normalization_gan.py]: The chance is quite small but I would deal with that

**Comment by sandeep-krishnamurthy** [comment]
Thanks for your contributions @stu1130 
@ThomasDelteil - Can you please take a look at this? Thank you.

**Comment by ThomasDelteil** [comment]
@stu1130 thanks for your contribution, can you add a README.md in the folder of the sn_gan with reference to papers / sample output etc.
Have a look at this one for inspiration for example: https://github.com/apache/incubator-mxnet/tree/master/example/gluon/embedding_learning

**Comment by vandanavk** [comment]
Please execute pylint on these files using ci/other/pylintrc in incubator-mxnet folder and fix errors (if any).

pylint --rcfile=ci/other/pylintrc --ignore-patterns="..so$$,..dll$$,..dylib$$"  example/gluon/sn-gan/*.py

File: example/gluon/sn-gan/spectral_normalization_gan.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+
+BATCH_SIZE = 64  # input batch size
+IMAGE_SIZE = 64  # image size
+Z_DIM = 100  # dimension of the latent z vector
+NUM_ITER = 1000  # number of epochs to train for
+LEARNING_RATE = 0.0002  # learning rate
+BETA = 0.5  # beta1 for adam
+OUTPUT_DIR = './data'  # output directory
+MANUAL_SEED = random.randint(1, 10000)  # manual seed
+CTX = mx.gpu()  # change to gpu if you have gpu
```
Comment (vandanavk) [inline example/gluon/sn-gan/spectral_normalization_gan.py]: Default should be cpu?

Comment (stu1130) [reply example/gluon/sn-gan/spectral_normalization_gan.py]: I think I should use argparse to let the user specify parameters

File: example/gluon/sn-gan/README.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
+# Spectral Normalization GAN
+
+This example implements [Spectral Normalization for Generative Adversarial Networks](https://openreview.net/pdf?id=B1QRgziT-) based on [CIFAR10](https://www.cs.toronto.edu/~kriz/cifar.html) dataset.
```
Comment (sandeep-krishnamurthy) [inline example/gluon/sn-gan/README.md]: Please refer to arxiv - https://arxiv.org/abs/1802.05957

File: example/gluon/sn-gan/README.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+## Usage
+
+Example runs and the results:
+
+```python
+python train.py --use-gpu --data-path=data/CIFAR10
```
Comment (sandeep-krishnamurthy) [inline example/gluon/sn-gan/README.md]: Write a note that user needs to download CIFAR10 dataset

File: example/gluon/sn-gan/model.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+EPSILON = 1e-08
+POWER_ITERATION = 1
+
+class SNConv2D(Block):
+    """ Customized Conv2D to feed the conv with the weight that we apply spectral normalization """
+
+    def __init__(self, num_filter, kernel_size,
+                 strides, padding, in_channels,
+                 ctx, iterations=1):
```
Comment (sandeep-krishnamurthy) [inline example/gluon/sn-gan/model.py]: set default ctx=mx.cpu()

File: example/gluon/sn-gan/model.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+            no_bias=True
+        )
+
+
+def get_generator():
+    """ construct and return generator """
+    g_net = gluon.nn.Sequential()
+    with g_net.name_scope():
+
+        g_net.add(gluon.nn.Conv2DTranspose(512, 4, 1, 0, use_bias=False))
```
Comment (sandeep-krishnamurthy) [inline example/gluon/sn-gan/model.py]: It is always more clear for readers in these kind of examples to have named parameters for these layers.

File: example/gluon/sn-gan/data.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+def transformer(data, label):
+    """ data preparation """
+    data = mx.image.imresize(data, IMAGE_SIZE, IMAGE_SIZE)
+    data = mx.nd.transpose(data, (2, 0, 1))
+    data = data.astype(np.float32) / 128.0 - 1
+    return data, label
+
+
+def get_training_data(batch_size):
+    """ hepler function to get dataloader"""
```
Comment (ThomasDelteil) [inline example/gluon/sn-gan/data.py]: helper

File: example/gluon/sn-gan/model.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+            # init the weight
+            self.weight = self.params.get('weight', shape=(
+                num_filter, in_channels, kernel_size, kernel_size))
+            self.u = self.params.get(
+                'u', init=mx.init.Normal(), shape=(1, num_filter))
+
+    def spectral_norm(self):
+        """ spectral normalization """
+        w = self.params.get('weight').data(self.ctx)
+        w_mat = w
```
Comment (ThomasDelteil) [inline example/gluon/sn-gan/model.py]: I don't think this is necessary, you can simply use `w` in your `nd.reshape(w_mat..`

File: example/gluon/sn-gan/model.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+        self.ctx = ctx
+
+        with self.name_scope():
+            # init the weight
+            self.weight = self.params.get('weight', shape=(
+                num_filter, in_channels, kernel_size, kernel_size))
+            self.u = self.params.get(
+                'u', init=mx.init.Normal(), shape=(1, num_filter))
+
+    def spectral_norm(self):
```
Comment (ThomasDelteil) [inline example/gluon/sn-gan/model.py]: I would suggest using `_spectral_norm(self)` as it is a private function

File: example/gluon/sn-gan/train.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+parser.add_argument('--data-path', type=str, default='./data',
+                    help='path of data.')
+parser.add_argument('--batch-size', type=int, default=64,
+                    help='training batch size. default is 64.')
+parser.add_argument('--epochs', type=int, default=100,
+                    help='number of training epochs. default is 100.')
+parser.add_argument('--lr', type=float, default=0.0001,
+                    help='learning rate. default is 0.0001.')
+parser.add_argument('--lr-beta', type=float, default=0.5,
+                    help='learning rate for the beta in margin based loss. default is 0.5s.')
```
Comment (ThomasDelteil) [inline example/gluon/sn-gan/train.py]: 0.5s ? what does `s` stand for here?

File: example/gluon/sn-gan/README.md [hunk 12]
```diff
@@ -0,0 +1,40 @@
+# Spectral Normalization GAN
```
Comment (ThomasDelteil) [inline example/gluon/sn-gan/README.md]: Please update  https://github.com/apache/incubator-mxnet/tree/master/example#deep-learning-examples-in-the-mxnet-project-repository with your example, thanks!

File: example/gluon/sn-gan/README.md [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+  --use-gpu             use gpu for training.
+  --clip_gr CLIP_GR     Clip the gradient by projecting onto the box. default
+                        is 10.0.
+  --z-dim Z_DIM         dimension of the latent z vector. default is 100.
+```
+
+## Learned Spectral Normalization
+
+![alt text](https://github.com/taki0112/Spectral_Normalization-Tensorflow/blob/master/assests/sn.png)
+
```
Comment (ThomasDelteil) [inline example/gluon/sn-gan/README.md]: Could you add a few samples of the generated images? It always makes it more appealing for people looking to try new models. Thanks!

Comment (stu1130) [reply example/gluon/sn-gan/README.md]: I am running the model with Xavier initializer and will update the image if it's better

File: example/gluon/sn_gan/model.py [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+            self.weight = self.params.get('weight', shape=(
+                num_filter, in_channels, kernel_size, kernel_size))
+            self.u = self.params.get(
+                'u', init=mx.init.Normal(), shape=(1, num_filter))
+
+    def _spectral_norm(self):
+        """ spectral normalization """
+        w = self.params.get('weight').data(self.ctx)
+        # the w preserve the original weight value to be used in line 75
+        w_mat = w
```
[the remark this answers is not in the document]
Comment (ThomasDelteil) [reply example/gluon/sn_gan/model.py]: assignment does not create a copy, reshape creates a copy. you can use:
`w_mat = nd.reshape(w, [w.shape[0], -1])`

File: example/gluon/sn_gan/README.md [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+optional arguments:
+  -h, --help            show this help message and exit
+  --data-path DATA_PATH
+                        path of data.
+  --batch-size BATCH_SIZE
+                        training batch size. default is 64.
+  --epochs EPOCHS       number of training epochs. default is 100.
+  --lr LR               learning rate. default is 0.0001.
+  --lr-beta LR_BETA     learning rate for the beta in margin based loss.
+                        default is 0.5s.
```
Comment (ThomasDelteil) [inline example/gluon/sn_gan/README.md]: nit: the `s` is also here still
